connect/runtime/src/test/java/org/apache/kafka/connect/runtime/TestConverterWithHeaders.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.Map;
+import org.apache.kafka.common.errors.SerializationException;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.storage.Converter;
+
+/**
+ * This is a simple Converter implementation that uses header values for serialization / deserialization purposes.
+ *
+ * It expects "message.type" header to be set with "MessageTypeA" or "MessageTypeB" value indicating target class for serialization / deserialization.
+ *
+ * To avoid using any dependencies like Avro or Protobuf it uses Java serialization.
```

Comment:
While I don't want to create more work, the use of Java serialization here gives me pause. Yes, this is only used within the tests themselves, but Java serialization has historically had numerous CVEs and I wonder if it's simply better to avoid Java serialization altogether in the AK codebase. Scanners are not terribly sophisticated, and it'd be a shame for this change to increase the burden of dealing with potential questions about AK and serialization.

A simpler test converter implementation might simply serialize strings to bytes using different encodings and/or use different artificial prefixes that the tests could assert never appear in the deserialized messages.

WDYT? Is there an advantage of the current use of Java serialization over an alternative (other than the fact it's complete)?

Reply:
FWIW, moving away from Java serialization would be a great follow-on PR, but I don't think it should delay merging this.

Reply:
@rhauch I mentioned the reason why I started with Java serialization (it's a "real" serialization, but no external libraries needed), but I can understand your concerns.

If you don't agree with @ryannedolan I can come up with something creative, like maybe using different compression algorithms (lz4, snappy, etc.) as serialization metadata instead (and just serialize simple strings), since Kafka comes with all those compression libraries anyway. WDYT?

Reply:
The point of the test converter is really just to ensure that the converter is being used correctly, right? Do we really need the test converter to do more than help verify that? I don’t think compression is really necessary. Something simple should work just fine. 

Because we are getting somewhat close to KIP freeze, I’d be concerned that a follow on PR fails to be merged in time. If we can keep this simple, it seems beneficial to change this PR.

Reply:
Got it! I'll use some of your suggestions.

Reply:
@rhauch simplified test converter to use string encoding in https://github.com/apache/kafka/pull/6362/commits/993966a68e9b6744364dece9ff7cba9fa77546b1

let me know if it looks good and I'll resolve merge conflicts.